rustls/src/client/hs.rs

```diff
@@ -300,8 +300,25 @@ fn emit_client_hello_for_retry(
 
     if let Some(key_share) = &key_share {
         debug_assert!(support_tls13);
-        let key_share = KeyShareEntry::new(key_share.group(), key_share.pub_key());
-        exts.push(ClientExtension::KeyShare(vec![key_share]));
+        let mut shares = vec![KeyShareEntry::new(key_share.group(), key_share.pub_key())];
+
+        if retryreq.is_none() {
+            // Only for the initial client hello, see if we can send a second KeyShare
+            // for "free".  We only do this if the same algorithm is also supported
+            // separately by our provider for this version (`find_kx_group` looks that up).
+            if let Some(hybrid_pair) = key_share.hybrid_pair().filter(|pair| {
+                config
+                    .find_kx_group(pair.group(), ProtocolVersion::TLSv1_3)
```

Comment:
It seems kind of a confusing that `pair` has a single `group()`? Shouldn't it conceptually have two, or something?